### ISSUE TITLE
server: add plugin support

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#13901]
tldr: We should add plugins. Plugins need to get a `client.DB`, but `client` is in the `internal` directory, so can't be passed to plugins. We should move `client` outside the `internal` directory, which is where it used to be.

---

I'd like to discuss if and how we should allow plugins. For reference, I have revived the redis branch using go plugins: https://github.com/cockroachdb/cockroach/compare/master...mjibson:redis-plugin

This works by adding a cmux Matcher in the server, and handing off connections and a `client.DB` to the plugin implementation:

https://github.com/mjibson/cockroach/commit/dbe2ce09a0c1f14c3f3668a13fec5a94bbf638ec#diff-09856fe9becddf0199651f451409356a

The plugin use there is just a POC and I do not wish to bikeshed its semantics here. The question I have here is regarding the handling of `client.DB`. Currently the client package is in the internal directory, so nothing outside cockroach can see it. As a work around I created the redis package in cockroach so it would have access to the client package, however this clearly isn't a general solution. If we decide to allow plugins, we need a better solution.

Assuming we are ok with the idea of plugins, it seems that they can have varying points of entry. We may end up supporting some or all of these. (I'm not discussing SQL plan hook plugins here because those would require modifying the SQL grammar also which is out of scope of this issue.) Here's the two I have come up with:

1. Add a cmux listener that comes after the pg listener and before the any listener. This has the risk of breaking grpc traffic if the plugin matcher is poorly defined. It does not help with regard to SSL because that is handled individually by each listener. This may make differentiating SSL-enabled redis and GRPC impossible, so may not be a viable solution.
1. Add a hook where the plugin can choose its own port to listen on.

In both of these cases, an underlying data store must be given to the plugin. `client.DB` is probably what we'll need, but we can't do that now because of the internal stuff listed above. I see two solutions:

1. Move `client` out of internal. It didn't used to be there, so maybe it's not so bad to move it back out.
1. Make a public interface of the important client methods: get/put/scan, txn, batches, etc. This seems pretty silly, though, since it'd just be a big copy of all the client stuff. We should only do this if there's a really good reason to leave client in internal.

Jira issue: CRDB-6109

**Comment by dt** [comment]
I'm apprehensive about adding features that aren't supported across OSs -- while production env's will likely always be Linux, plenty of devs will likely evaluate first on their macs. Should we wait for Darwin plugin support to land later this year?

**Comment by petermattis** [comment]
The proposal only covers a single form of plugin: allowing a plugin to provide a new user API. There are lots of other potential hooks, such as adding new SQL functions, types or virtual tables. Plugins bring an additional set of headaches. There would need to be administrative commands for listing the active plugins, disabling plugins, etc. How would we test plugins for compatibility? What compatibility requirements would there be between the plugin and the version of CockroachDB?

There are likely answers to these questions, but adding support for plugins definitely seems to fall in the Later milestone.

**Comment by bdarnell** [comment]
-1 on allowing plugins to add matchers to the main cmux listener. This is already fragile enough (and we've been unable to multiplex our web interface onto this port). I'd rather require that any plugin that exposes a new network interface do so on an new port. With that restriction, does it even need to be a plugin instead of a separate process that talks to the local cockroachdb node over its regular network interface (either sql or grpc)? This is similar to the bigtable/spanner coprocessor model and would allow these coprocessors to be written in any language, not just go (and more easily supported cross-platform, etc). For other kinds of plugins (new sql functions, etc), the argument for linking them into the main process is stronger, but it's still riskier to do so than to hold the extension at arm's length. 

I'd also rather not have plugins writing directly to `client.Txn`. We made that `internal` because it's an internal interface with lots of sharp edges and limited compatibility guarantees. We should eat our own dogfood and use the public sql interface unless there's a really strong need for a lower-level interface.

**Comment by dt** [comment]
I'm not convinced about `internal` -- you can always just write and compile your plugin in a cockroach clone.

That said, where we allow plugins to hook in or not does seem interesting -- planHook seems like a great place to offer plugins the chance to jump in and handle additional SQL statements, and any SQL execution will certainly will want access to client.Txn/client.DB.

**Comment by bdarnell** [comment]
I'd be OK with renaming `internal/client` to something like `internalclient` and giving up the compiler enforcement if we do decide to expose this to plugins. The reason we put it in `internal` is that otherwise `cockroach/pkg/client` really looks like something intended for external consumption.

**Comment by mjibson** [comment]
I like many of the points you make about coprocessors, even though they would require people to run more monitoring and deployment infra around another service. But since plugins won't make it in any time soon, that seems like a good tradeoff. Being able to rename client to internalclient would make these coprocessors trivial to do in Go, which is great.

Renaming just the directory `internal/client` to `internalclient`, but leaving the `package client` name the same would allow (I think) the code to remain the same since the package name is the same, just the import paths changing. Any objections to this? It kind of stinks because the directory name and package name wouldn't be the same.

**Comment by bdarnell** [comment]
What does `goimports` think about the directory name and package name not being the same? Even if that works, I think I'd slightly prefer to rename all the references to keep the package and directory names the same.

To be clear, "coprocessors" are more than a completely independent service that the application would manage. The idea is that cockroach would manage the deployment and run a copy of the coprocessor on each node that has a replica of a table that uses the coprocessor. See [hbase's docs](https://blogs.apache.org/hbase/entry/coprocessor_introduction) on coprocessors.

**Comment by dt** [comment]
would just moving `client` as-is under `kv` be enough for appearances? it seems like getting a `Sender` is non-trivial enough i doubt you'd see any accidental usages.

**Comment by bdarnell** [comment]
I think `kv/client` is even worse since that's exactly what a lot of NoSQL fans are looking for. :)

**Comment by knz** [comment]
kv/client is a thing now.
